log-replication/src/main/java/org/corfudb/logreplication/fsm/LogReplicationFSM.java

```
@@ -0,0 +1,352 @@
+        states.put(LogReplicationStateType.STOPPED, new StoppedState());
+    }
+
+    /**
+     * Input function of the FSM.
+     *
+     * This method enqueues log replication events for further processing.
+     *
+     * @param event LogReplicationEvent to process.
+     */
+    public synchronized void input(LogReplicationEvent event) {
+        try {
+            if (state.getType().equals(LogReplicationStateType.STOPPED)) {
+                // Log: not accepting events, in stopped state
+                return;
+            }
+            if (event.getType() != LogReplicationEventType.LOG_ENTRY_SYNC_CONTINUE) {
+                log.info("Enqueue event {} with ID {}", event.getType(), event.getEventID());
+            }
+            eventQueue.put(event);
+        } catch (InterruptedException ex) {
+            log.error("Log Replication interrupted Exception: ", ex);
+        }
+    }
+
+    /**
+     * Consumer of the eventQueue.
+     *
+     * This method consumes the log replication events and does the state transition.
+     */
+    private void consume() {
+        try {
+            if (state.getType() == LogReplicationStateType.STOPPED) {
+                log.info("Log Replication State Machine has been stopped. No more events will be processed.");
+                return;
+            }
+
+            // TODO (Anny): consider strategy for continuously failing snapshot sync (never ending cancellation)
+            //   Block until an event shows up in the queue.
+            LogReplicationEvent event = eventQueue.take();
+            if (event.getType() != LogReplicationEventType.LOG_ENTRY_SYNC_CONTINUE) {
+                log.info("Log Replication FSM consume event {}", event);
+            }
+
+            if (event.getType() == LogReplicationEventType.LOG_ENTRY_SYNC_REPLICATED) {
+                if (state.getType() == LogReplicationStateType.IN_LOG_ENTRY_SYNC &&
```

we should add an assert or throw an exception if this event was received from an unexpected state.